### PR TITLE
Define an Atelier service-tier proposal and pilot migration plan

### PR DESCRIPTION
**Pull request by shaug**
# Summary

- Define a service/use-case tier proposal for Atelier workflows that orchestrate
  multiple side effects.
- Capture a bounded pilot migration plan before broader refactoring.

# Changes

- Added `docs/service-tier-proposal.md` with:
  - layer boundaries, naming conventions, and module mapping from current code
  - result/error contract guidance for side-effect-heavy workflows
  - adoption guardrails, anti-goals, and phased pilot exit criteria
- Incorporated review feedback to clarify:
  - dependency injection seams for collaborators and request context
  - composition-depth guidance with a documented exception path
  - typing-first validation plus explicit constraint validation
  - when to return `ServiceFailure` vs when to raise exceptions
  - observability baseline expectations for service entrypoints
- Updated `docs/worker-runtime-architecture.md` to link to the proposal.

# Testing

- `just format` ✅
- `just lint` ✅
- `just test` ❌ (existing collection failure on Python 3.14:
  `importlib.abc.Traversable` import in `src/atelier/skills.py`)

# Risks / Rollout

- Docs-only changes; no runtime behavior changes.
- Proposal acts as a north-star for the next bounded pilot refactor
  (`atelier init` + config prompting path).

### CODE REVIEW

File: docs/service-tier-proposal.md
Comment (shaug) [inline docs/service-tier-proposal.md]: Suggested change: add an “Inspiration” section with link to https://github.com/sunny/actor, and how this proposal matches its intent, and any significant differences (esp for a Python vs Ruby project)

File: docs/service-tier-proposal.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+- `RunStartupContractService`
+  - wraps startup selection flow now split between `worker/session/runner.py`,
+    `worker/session/startup.py`, and `worker/work_startup_runtime.py`
+- `ComposeWorkerOpeningPromptService`
+  - wraps prompt assembly and mode-specific policy text
+- `FinalizeChangesetService`
+  - wraps finalization/publish decision orchestration while preserving existing
+    pipeline helpers
+
+## Result and error contracts
```
Comment (shaug) [inline docs/service-tier-proposal.md]: Proposal doesn’t define where validation must happen in the service lifecycle.

Why it matters: a core concept should be validated input contracts before business logic, so that inputs are considered valid *before* business logic is applied

Suggested change: add a rule like “each service entrypoint validates request/context at boundary and returns validation_failed deterministically.” Strongly recommend relying on strong typing and pydantic for defining schemas when inputs must be validated from external sources

File: docs/service-tier-proposal.md
Comment (shaug) [reply docs/service-tier-proposal.md]: Added in commit `fb93ab4`.

The proposal now includes a dedicated **Inspiration and alignment** section that links ServiceActor and explicitly calls out alignment plus key Python-specific differences for Atelier.

File: docs/service-tier-proposal.md (hunk 1 above)
Comment (shaug) [reply docs/service-tier-proposal.md]: Addressed in `fb93ab4`.

I added an explicit validation rule under result/error contracts: each service validates request/context at the boundary before orchestration, returns `validation_failed` deterministically, and prefers Pydantic for untrusted external inputs.

File: docs/service-tier-proposal.md [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+Use explicit tagged outcomes for side-effect-heavy services.
+
+Validation rule:
+
+- Every service entrypoint validates request/context input before business
+  orchestration starts.
+- Validation failures must return `validation_failed` deterministically with a
```
[the remark this answers is not in the document]
Comment (shaug) [reply docs/service-tier-proposal.md]: Documented now. The contract section explicitly splits failure handling into:
- `ServiceFailure` for expected domain/policy/runtime failures the caller can handle deterministically
- raised exceptions for programmer bugs/invariant violations that indicate defects

File: docs/service-tier-proposal.md [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+- decide whether to extend pattern to worker startup and/or finalization
+
+Exit criteria:
+
+- written decision record for expand/adjust/stop
+- if expanding, define next bounded changeset per workflow
+
+## Success metrics for pilot review
+
+- lower orchestration complexity in command/runtime entry modules
```
[the remark this answers is not in the document]
Comment (shaug) [reply docs/service-tier-proposal.md]: Added to success metrics. The pilot now includes an observability baseline requiring structured start/success/failure logs at service entrypoints plus duration measurement.

File: docs/service-tier-proposal.md [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+- success outcome
+  - lifecycle action taken (`pushed`, `pr-open`, `merged`, etc.)
+  - resulting refs/PR URL when available
+- failure codes
+  - `policy_blocked` when PR strategy gates creation
+  - `external_command_failed` when git/gh commands fail
+  - `unexpected_state` for inconsistent changeset metadata
+
+## Adoption guardrails
```
Comment (shaug) [inline docs/service-tier-proposal.md]: Could we add a short decision checklist under this section for “service vs helper” classification (for example: side effects? policy gates? stable failure contract needed? orchestration across adapters?). This would reduce interpretation drift as this doc becomes the north-star.

File: docs/service-tier-proposal.md [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+        "io_failed",
+        "unexpected_state",
+    ]
+    message: str
+    recovery_hint: str | None = None
+
+ServiceResult = ServiceSuccess[T] | ServiceFailure
+```
+
+### Contract example: init/config orchestration
```
Comment (shaug) [inline docs/service-tier-proposal.md]: Could we add one compact “golden path” example here that shows the full flow end-to-end (controller parses input -> service request/context -> `ServiceResult` success/failure handling)? It would make the contract concrete for future contributors implementing CS2.

File: docs/service-tier-proposal.md [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+- this proposal document
+- module mapping and guardrails
+- typed result/error contract examples
+
+Exit criteria:
+
+- maintainers agree on service naming and boundary model
+- pilot path is explicitly scoped before code movement
+
+### Phase CS2: bounded init/config prompt refactor
```
Comment (shaug) [inline docs/service-tier-proposal.md]: To make CS2 verification more deterministic, can we add a brief “service test levels” rubric here (service unit tests with injected fakes, contract tests for request/response boundaries, and command-level integration tests)? That gives a clear definition of done for pilot quality.

File: docs/service-tier-proposal.md (hunk 4 above)
Comment (shaug) [reply docs/service-tier-proposal.md]: Added in the proposal.

Under **Adoption guardrails**, there is now a **Service vs helper decision checklist** with the requested criteria:
- side-effect boundary coordination
- policy/lifecycle gates
- stable failure-code dependency
- orchestration across adapters/ports

It also includes a concrete threshold rule ("two or more yes" => prefer a service entrypoint) to reduce interpretation drift.

File: docs/service-tier-proposal.md (hunk 5 above)
Comment (shaug) [reply docs/service-tier-proposal.md]: Added a compact end-to-end golden path in the contracts section.

The new **Golden path example (controller to service result handling)** shows:
- controller input parsing
- construction of a typed service request
- `InitializeProjectService.run(...)`
- deterministic `ServiceSuccess`/`ServiceFailure` handling and exit mapping

This is intended to be the concrete CS2 reference flow.

File: docs/service-tier-proposal.md (hunk 6 above)
Comment (shaug) [reply docs/service-tier-proposal.md]: Added this in the CS2 section.

There is now a **Service test levels rubric** covering:
- service unit tests with injected fakes
- request/response contract tests for validation and stable error codes
- command-level integration tests for controller wiring and exit/render compatibility

This should make CS2 quality expectations and verification scope deterministic.
